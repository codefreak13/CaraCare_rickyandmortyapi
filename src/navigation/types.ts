import {RouteProp} from '@react-navigation/native';
import {NativeStackNavigationProp} from '@react-navigation/native-stack';
import {Character} from '../apollo/query';

/// We use enums to prevent the use of strings across the application - i.e, write once

export const enum APP_ROUTE {
  LIST = 'LIST',
  FAVORITE = 'FAVORITE',
  DETAIL = 'DETAIL',
}

export type AppStackParamList = {
  [APP_ROUTE.LIST]: undefined;
  [APP_ROUTE.FAVORITE]: undefined;
  [APP_ROUTE.DETAIL]: Character;
};

export type ListScreenNavigationProp = NativeStackNavigationProp<
  AppStackParamList,
  APP_ROUTE.LIST
>;

export type FavoriteScreenNavigationProp = NativeStackNavigationProp<
  AppStackParamList,
  APP_ROUTE.FAVORITE
>;

export type DetailScreenRouteProp = RouteProp<
  AppStackParamList,
  APP_ROUTE.DETAIL
>;

export type DetailScreenNavigationProp = NativeStackNavigationProp<
  AppStackParamList,
  APP_ROUTE.DETAIL
>;
