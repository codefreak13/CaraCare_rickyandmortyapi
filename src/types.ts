interface Episode {
  name: string;
}

export interface Character {
  id: string;
  name: string;
  status: string;
  species: string;
  gender: string;
  origin: {
    name: string;
  };
  episode: Episode[];
  created: string;
  image: string;
}

interface FilterCharacter {
  status?: string;
}

interface Info {
  count: number;
  pages: number;
  next: number;
  prev: number;
}

export interface CharacterVars {
  filter?: FilterCharacter;
  page?: number;
}

export interface CharacterData {
  characters: {
    info: Info;
    results: Character[];
  };
}

export const API = 'https://rickandmortyapi.com/graphql';

export enum FLEX_TYPE {
  columnReverse = 'column-reverse',
  row = 'row',
}

export enum WRAP {
  wrap = 'wrap',
  nowrap = 'nowrap',
  wrapReverse = 'wrap-reverse',
}

export enum ALIGN_TYPE {
  center = 'center',
  flexStart = 'flex-start',
}

export enum VIEW_TYPE {
  list = 'list',
  grid = 'grid',
}

export enum FAV_ICON {
  heart = 'heart',
  heartOutline = 'heart-outline',
}

export enum POSITION_TYPE {
  absolute = 'absolute',
  relative = 'relative',
}

export enum SIZE {
  small = 'small',
  large = 'large',
}

export type ListItemProps = {
  viewType?: string;
  onPress?: () => void;
  addToFavorite?: () => void;
} & Character;

export const enum STORAGE_VALUES {
  FAVORITE_LIST = 'FAVORITE_LIST',
  VIEW_TYPE = 'VIEW_TYPE',
}
