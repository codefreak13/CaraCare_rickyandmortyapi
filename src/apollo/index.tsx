import {ApolloClient, HttpLink, InMemoryCache} from '@apollo/client';

export const client = new ApolloClient({
  link: new HttpLink({uri: 'http://rickandmortyapi.com/graphql'}),
  //   uri: 'http://rickandmortyapi.com/graphql',
  cache: new InMemoryCache({
    typePolicies: {
      Query: {
        fields: {
          characters: {
            keyArgs: false,
            merge(existing, incoming) {
              if (!existing) return incoming;
              const newValue = {...existing};
              newValue.info = incoming.info;
              newValue.results = [...existing.results, ...incoming.results];
              return newValue;
            },
          },
        },
      },
    },
  }),
});
