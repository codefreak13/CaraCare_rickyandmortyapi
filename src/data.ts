const statusData = [
  {label: 'Alive', value: 'Alive'},
  {label: 'Dead', value: 'Dead'},
];

export {statusData};
