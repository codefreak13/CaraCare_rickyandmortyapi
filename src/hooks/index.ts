export {default as useList} from './useList';
export {default as useListItem} from './useListItem';
export {default as useDetail} from './useDetail';
export {default as useFavorite} from './useFavorite';
export {default as useContextHook} from './useContextHook';
