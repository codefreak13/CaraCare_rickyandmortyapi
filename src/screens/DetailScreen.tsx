import React from 'react';
import {Header} from '../components';
import {useDetail} from '../hooks';
import {BasicListItem} from '../ui';

const DetailScreen = () => {
  const {
    numOfEpisodes,
    firstEpisode,
    lastEpisode,
    customStyle,
    originName,
    imageStyle,
    species,
    status,
    gender,
    goBack,
    image,
    name,
  } = useDetail();

  return (
    <>
      <Header
        leftButton={{
          onclick: () => goBack(),
        }}
        customMiddleIcon
        title="CHARACTER DETAIL"
      />
      <BasicListItem
        name={name}
        image={image}
        status={status}
        gender={gender}
        species={species}
        origin={originName}
        episode={firstEpisode}
        imageStyle={imageStyle}
        lastEpisode={lastEpisode}
        customStyle={customStyle}
        numOfEpisodes={numOfEpisodes}
      />
    </>
  );
};

export default DetailScreen;
