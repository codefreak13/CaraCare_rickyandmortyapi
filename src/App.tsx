import React from 'react';
import {client} from './apollo';
import {ApolloProvider} from '@apollo/client';
import AppContextProvider from './store/Context';
import Screens from './navigation/RootNavigator';
import {SafeAreaView, StatusBar, StyleSheet} from 'react-native';

const App = () => {
  return (
    <SafeAreaView style={styles.backgroundStyle}>
      <StatusBar barStyle={'light-content'} />
      <ApolloProvider client={client}>
        <AppContextProvider>
          <Screens />
        </AppContextProvider>
      </ApolloProvider>
    </SafeAreaView>
  );
};

export default App;

const styles = StyleSheet.create({
  backgroundStyle: {
    flex: 1,
  },
});
