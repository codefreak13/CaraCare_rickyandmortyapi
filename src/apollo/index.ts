import {ApolloClient, HttpLink, InMemoryCache} from '@apollo/client';
import {API} from 'src/types';

export const client = new ApolloClient({
  link: new HttpLink({uri: API}),
  cache: new InMemoryCache({
    typePolicies: {
      Query: {
        fields: {
          characters: {
            // Don't cache separate results based on
            // any of this field's arguments.
            keyArgs: false,
            // Concatenate the incoming list items with
            // the existing list items.
            merge(existing, incoming) {
              if (!existing) return incoming;
              const newValue = {...existing};
              newValue.info = incoming.info;
              newValue.results = [...existing.results, ...incoming.results];
              return newValue;
            },
          },
        },
      },
    },
  }),
});
