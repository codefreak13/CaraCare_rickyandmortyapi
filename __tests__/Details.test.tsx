import React from 'react';
import {mockData} from 'src/data';
import DetailScreen from 'src/screens/DetailScreen';
import AppContextProvider from 'src/store/Context';
import {MockedProvider} from '@apollo/client/testing';
import {NavigationContainer} from '@react-navigation/native';
import {render} from '@testing-library/react-native';
import {AppStackParamList, APP_ROUTE} from 'src/navigation/types';
import {createNativeStackNavigator} from '@react-navigation/native-stack';
import {Character} from 'src/types';

const Stack = createNativeStackNavigator<
  AppStackParamList & {MOCK_COMPONENT: undefined}
>();

describe('DetailScreen', () => {
  it('renders DetailScreen correctly', async () => {
    const {toJSON, getByText} = render(
      <MockedProvider mocks={mockData}>
        <AppContextProvider>
          <NavigationContainer>
            <Stack.Navigator screenOptions={{headerShown: false}}>
              <Stack.Screen
                name={APP_ROUTE.DETAIL}
                component={DetailScreen}
                initialParams={
                  mockData[0].data.results[0] as Partial<Character>
                }
              />
            </Stack.Navigator>
          </NavigationContainer>
        </AppContextProvider>
      </MockedProvider>,
    );

    expect(toJSON()).toMatchSnapshot();
    expect(getByText('CHARACTER DETAIL')).toBeTruthy();
  });
});
