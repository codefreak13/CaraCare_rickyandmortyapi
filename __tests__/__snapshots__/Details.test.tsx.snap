// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailScreen renders DetailScreen correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RNSScreenStack
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <RNSScreen
      collapsable={false}
      forwardedRef={[Function]}
      gestureResponseDistance={
        Object {
          "bottom": -1,
          "end": -1,
          "start": -1,
          "top": -1,
        }
      }
      onAppear={[Function]}
      onDisappear={[Function]}
      onDismissed={[Function]}
      onTransitionProgress={[Function]}
      onWillDisappear={[Function]}
      replaceAnimation="push"
      stackPresentation="push"
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      swipeDirection="horizontal"
    >
      <RNSScreenStackHeaderConfig
        backButtonInCustomView={false}
        backgroundColor="rgb(255, 255, 255)"
        color="rgb(0, 122, 255)"
        direction="ltr"
        disableBackButtonMenu={false}
        hidden={true}
        hideBackButton={false}
        largeTitleHideShadow={false}
        title="DETAIL"
        titleColor="rgb(28, 28, 30)"
        topInsetEnabled={false}
        translucent={false}
      />
      <View
        style={
          Array [
            Object {
              "flex": 1,
            },
            Object {
              "backgroundColor": "rgb(242, 242, 242)",
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "height": 83.375,
                "width": 750,
              },
              null,
              null,
            ]
          }
        >
          <View
            style={
              Object {
                "left": 15,
                "position": "absolute",
                "zIndex": 1,
              }
            }
          >
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <Text
                allowFontScaling={false}
                selectable={false}
                style={
                  Array [
                    Object {
                      "color": "#1D1F20",
                      "fontSize": 16,
                    },
                    undefined,
                    Object {
                      "fontFamily": "FontAwesome5Free-Solid",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {
                      "fontWeight": "900",
                    },
                  ]
                }
              >
                ÔÅ†
              </Text>
            </View>
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "bottom": 0,
                "justifyContent": "center",
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "fontFamily": undefined,
                    "fontSize": 26.5,
                    "fontWeight": "bold",
                  },
                  Object {
                    "color": "#1D1F20",
                  },
                ]
              }
            >
              CHARACTER DETAIL
            </Text>
          </View>
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "backgroundColor": "#FFF",
                "flex": 1,
                "flexDirection": "row",
                "margin": 1.5,
                "marginVertical": 8,
                "padding": 11.5,
                "shadowColor": "grey",
                "shadowOffset": Object {
                  "height": 3.5,
                  "width": 8,
                },
                "shadowOpacity": 1,
                "zIndex": 999,
              },
              Object {
                "alignItems": "center",
                "flexDirection": "row",
              },
            ]
          }
        >
          <View
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "flexDirection": "row",
                    "flexWrap": "wrap",
                  },
                  undefined,
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": undefined,
                      "fontSize": 26.5,
                      "fontWeight": "bold",
                    },
                    Object {
                      "alignSelf": "center",
                      "backgroundColor": "#C3C9CE",
                      "color": "#1D1F20",
                      "marginBottom": 1.5,
                      "marginRight": 8,
                      "minWidth": 107,
                      "padding": 3.5,
                    },
                  ]
                }
              >
                Name
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": undefined,
                      "fontSize": 26.5,
                      "fontWeight": undefined,
                    },
                    Object {
                      "color": "#1D1F20",
                      "flexWrap": "wrap",
                      "maxWidth": 246.5,
                    },
                  ]
                }
                testID="itemName"
              >
                Mark
              </Text>
            </View>
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "flexDirection": "row",
                    "flexWrap": "wrap",
                  },
                  undefined,
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": undefined,
                      "fontSize": 26.5,
                      "fontWeight": "bold",
                    },
                    Object {
                      "alignSelf": "center",
                      "backgroundColor": "#C3C9CE",
                      "color": "#1D1F20",
                      "marginBottom": 1.5,
                      "marginRight": 8,
                      "minWidth": 107,
                      "padding": 3.5,
                    },
                  ]
                }
              >
                Status
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": undefined,
                      "fontSize": 26.5,
                      "fontWeight": undefined,
                    },
                    Object {
                      "color": "#1D1F20",
                      "flexWrap": "wrap",
                      "maxWidth": 246.5,
                    },
                  ]
                }
              >
                Alive
              </Text>
            </View>
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "flexDirection": "row",
                    "flexWrap": "wrap",
                  },
                  undefined,
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": undefined,
                      "fontSize": 26.5,
                      "fontWeight": "bold",
                    },
                    Object {
                      "alignSelf": "center",
                      "backgroundColor": "#C3C9CE",
                      "color": "#1D1F20",
                      "marginBottom": 1.5,
                      "marginRight": 8,
                      "minWidth": 107,
                      "padding": 3.5,
                    },
                  ]
                }
              >
                Species
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": undefined,
                      "fontSize": 26.5,
                      "fontWeight": undefined,
                    },
                    Object {
                      "color": "#1D1F20",
                      "flexWrap": "wrap",
                      "maxWidth": 246.5,
                    },
                  ]
                }
              >
                human
              </Text>
            </View>
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "flexDirection": "row",
                    "flexWrap": "wrap",
                  },
                  undefined,
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": undefined,
                      "fontSize": 26.5,
                      "fontWeight": "bold",
                    },
                    Object {
                      "alignSelf": "center",
                      "backgroundColor": "#C3C9CE",
                      "color": "#1D1F20",
                      "marginBottom": 1.5,
                      "marginRight": 8,
                      "minWidth": 107,
                      "padding": 3.5,
                    },
                  ]
                }
              >
                Gender
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": undefined,
                      "fontSize": 26.5,
                      "fontWeight": undefined,
                    },
                    Object {
                      "color": "#1D1F20",
                      "flexWrap": "wrap",
                      "maxWidth": 246.5,
                    },
                  ]
                }
              >
                Male
              </Text>
            </View>
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "flexDirection": "row",
                    "flexWrap": "wrap",
                  },
                  undefined,
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": undefined,
                      "fontSize": 26.5,
                      "fontWeight": "bold",
                    },
                    Object {
                      "alignSelf": "center",
                      "backgroundColor": "#C3C9CE",
                      "color": "#1D1F20",
                      "marginBottom": 1.5,
                      "marginRight": 8,
                      "minWidth": 107,
                      "padding": 3.5,
                    },
                  ]
                }
              >
                Origin
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": undefined,
                      "fontSize": 26.5,
                      "fontWeight": undefined,
                    },
                    Object {
                      "color": "#1D1F20",
                      "flexWrap": "wrap",
                      "maxWidth": 246.5,
                    },
                  ]
                }
              >
                Earth
              </Text>
            </View>
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "flexDirection": "row",
                    "flexWrap": "wrap",
                  },
                  undefined,
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": undefined,
                      "fontSize": 26.5,
                      "fontWeight": "bold",
                    },
                    Object {
                      "alignSelf": "center",
                      "backgroundColor": "#C3C9CE",
                      "color": "#1D1F20",
                      "marginBottom": 1.5,
                      "marginRight": 8,
                      "minWidth": 107,
                      "padding": 3.5,
                    },
                  ]
                }
              >
                First Episode
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": undefined,
                      "fontSize": 26.5,
                      "fontWeight": undefined,
                    },
                    Object {
                      "color": "#1D1F20",
                      "flexWrap": "wrap",
                      "maxWidth": 246.5,
                    },
                  ]
                }
              >
                Pilot
              </Text>
            </View>
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "flexDirection": "row",
                    "flexWrap": "wrap",
                  },
                  undefined,
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": undefined,
                      "fontSize": 26.5,
                      "fontWeight": "bold",
                    },
                    Object {
                      "alignSelf": "center",
                      "backgroundColor": "#C3C9CE",
                      "color": "#1D1F20",
                      "marginBottom": 1.5,
                      "marginRight": 8,
                      "minWidth": 107,
                      "padding": 3.5,
                    },
                  ]
                }
              >
                Last Episode
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": undefined,
                      "fontSize": 26.5,
                      "fontWeight": undefined,
                    },
                    Object {
                      "color": "#1D1F20",
                      "flexWrap": "wrap",
                      "maxWidth": 246.5,
                    },
                  ]
                }
              >
                Pilot
              </Text>
            </View>
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "flexDirection": "row",
                    "flexWrap": "wrap",
                  },
                  undefined,
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": undefined,
                      "fontSize": 26.5,
                      "fontWeight": "bold",
                    },
                    Object {
                      "alignSelf": "center",
                      "backgroundColor": "#C3C9CE",
                      "color": "#1D1F20",
                      "marginBottom": 1.5,
                      "marginRight": 8,
                      "minWidth": 107,
                      "padding": 3.5,
                    },
                  ]
                }
              >
                Number of Episodes
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": undefined,
                      "fontSize": 26.5,
                      "fontWeight": undefined,
                    },
                    Object {
                      "color": "#1D1F20",
                      "flexWrap": "wrap",
                      "maxWidth": 246.5,
                    },
                  ]
                }
              >
                1
              </Text>
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "overflow": "hidden",
                },
                Array [
                  Object {
                    "borderRadius": 82,
                    "height": 164.5,
                    "marginVertical": 24.5,
                    "width": 164.5,
                  },
                  Object {
                    "borderRadius": 246.5,
                    "height": 328.5,
                    "width": 328.5,
                  },
                ],
              ]
            }
          >
            <FastImageView
              resizeMode="cover"
              source={
                Object {
                  "uri": undefined,
                }
              }
              style={
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                }
              }
            />
          </View>
        </View>
      </View>
    </RNSScreen>
  </RNSScreenStack>
</RNCSafeAreaProvider>
`;
